str0m/src/session.rs

```diff
@@ -625,25 +641,58 @@ impl Session {
     }
 
     fn poll_packet(&mut self, now: Instant) -> Option<DatagramSend> {
+        self.update_queue_states(now);
+
         let srtp_tx = self.srtp_tx.as_mut()?;
+        // Only temporary to measure actual queuing delay for BWE work.
+        let real_now = Instant::now();
 
+        // Figure out which, if any, queue to poll
+        let PollOutcome::PollQueue(queue) = self.pacer.poll_action(now, &self.queue_states) else {
+            return None;
+        };
+
+        // NB: Cannot use mline_index_mut here due to borrowing woes around self, need split
+        // borrowing.
+        let mline = self.m_lines[queue.id.as_usize()]
+            .as_m_line_mut()
+            .expect("index is MLine");
         let buf = &mut self.poll_packet_buf;
 
-        for m in only_m_line_mut(&mut self.m_lines) {
-            let twcc_seq = self.twcc;
-            if let Some((header, seq_no)) = m.poll_packet(now, &self.exts, &mut self.twcc, buf) {
-                trace!("Poll RTP: {:?}", header);
-                let protected = srtp_tx.protect_rtp(buf, &header, *seq_no);
-                self.twcc_tx_register
-                    .register_seq(twcc_seq.into(), now, protected.len());
-
-                // const EGRESS_PACKET_LOSS_PERCENT: u64 = 5;
-                // if twcc_seq % (100 / EGRESS_PACKET_LOSS_PERCENT) == 0 {
-                //     return None;
-                // }
-
-                return Some(protected.into());
+        let twcc_seq = self.twcc;
+        if let Some((header, seq_no, queued_at)) =
+            mline.poll_packet(now, &self.exts, &mut self.twcc, buf)
+        {
+            if let Some(queued_at) = queued_at {
+                let delta_us = if real_now < queued_at {
+                    // Out of order
+                    let diff = queued_at
+                        .checked_duration_since(real_now)
+                        .map(|d| -(d.as_micros() as i64));
+                    warn!(?diff, "Packet sent before it was queued, clock drift?");
+                    diff.unwrap()
+                } else {
+                    real_now
+                        .checked_duration_since(queued_at)
+                        .map(|d| d.as_micros() as i64)
+                        .unwrap()
+                };
+                let kind = if mline.kind() == MediaKind::Audio {
+                    "audio"
+                } else {
+                    "video"
+                };
+                println!("QUEUE_STAT {},{},{}", header.ssrc, delta_us, kind);
```

Comment:
use debug instead of println

Reply:
This is temporary to get this output to STDOUT, I wrote a small python script to summarise these stats.

<details>
<summary>queue-latency.py</summary>

```python
import sys
import pandas as pd
import numpy as np
import io


def read_stats(path):
    output = io.StringIO()
    with open(path) as f:
        for line in f:
            if not line.startswith("QUEUE_STAT"):
                continue

            clean = line.removeprefix("QUEUE_STAT").strip()
            print(clean, file=output)

    output.seek(0)
    return output


def percentile(n):
    def percentile_(x):
        return np.percentile(x, n)

    percentile_.__name__ = "percentile_%s" % n
    return percentile_


if __name__ == "__main__":
    path = sys.argv[1]
    stats = read_stats(path)

    df = pd.read_csv(
        stats, header=None, names=["SSRC", "queue_time", "kind"], on_bad_lines="error"
    )
    queue_counts = df["queue_time"].value_counts()
    any_bad_values = False
    # df.drop(columns=["kind"], inplace=True)

    if any_bad_values:
        count = queue_counts[-1]
        print(f"Got {count} bad values for `queue_time`")

    # df = df.head(3000)
    # assert df.shape[0] == 3000

    stats = (
        df.groupby(["SSRC", "kind"])
        .agg(
            [
                percentile(50),
                percentile(75),
                percentile(95),
                percentile(99),
                "max",
                "count",
            ]
        )
        .reset_index()
    )

    print(stats)
```

</details>

I'll remove this before I merge but maybe it's something we want long term, it was useful for this work.

Reply:
nice

Reply:
Removed via rebase now.